File: java/src/P011/ContainerWithMostWater/ContainerWithMostWater.java
package P011.ContainerWithMostWater;

/*
Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

Note: You may not slant the container.
 */

public class ContainerWithMostWater {
    public static int maxArea(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }

        int area = 0;
        for (int i = 0, j = height.length - 1; i < j;) {
            int h = Math.min(height[i], height[j]);
            area = Math.max(area, (j - i) * h);
            while (height[i] <= h && i < j) i++;
            while (height[j] <= h && i < j) j--;
        }

        return area;
    }

    public static void main(String[] args) {
        int[] height = new int[] {1, 2, 3, 4};
        System.out.println(maxArea(height));
    }
}

File: python/P011_Container_With_Most_Water/ContainerWithMostWater.py
class Solution:
    @staticmethod
    def maxArea(height):
        if height is None or len(height) == 0:
            return 0
        i, j = 0, len(height) - 1
        area = 0
        while i < j:
            h = min(height[i], height[j])
            area = max(area, h * (j - i))
            while height[i] <= h and i < j:
                i += 1
            while height[j] <= h and i < j:
                j -= 1
        return area